lib/migration_priv.h

```diff
@@ -33,93 +33,28 @@
 
 #include <linux/vfio.h>
 
-/*
- * FSM to simplify saving device state.
- */
-enum migr_iter_state {
-    VFIO_USER_MIGR_ITER_STATE_INITIAL,
-    VFIO_USER_MIGR_ITER_STATE_STARTED,
-    VFIO_USER_MIGR_ITER_STATE_DATA_PREPARED,
-    VFIO_USER_MIGR_ITER_STATE_FINISHED
-};
-
 struct migration {
-    /*
-     * TODO if the user provides an FD then should mmap it and use the migration
-     * registers in the file
-     */
-    struct vfio_user_migration_info info;
+    uint64_t flags;
+    enum vfio_device_mig_state state;
     size_t pgsize;
     vfu_migration_callbacks_t callbacks;
-    uint64_t data_offset;
-
-    /*
-     * This is only for the saving state. The resuming state is simpler so we
-     * don't need it.
-     */
-    struct {
-        enum migr_iter_state state;
-        uint64_t pending_bytes;
-        uint64_t offset;
-        uint64_t size;
-    } iter;
 };
 
-struct migr_state_data {
-    uint32_t state;
-    const char *name;
+/* valid migration state transitions 
+   indexed by vfio_device_mig_state enum */
+static const bool transitions[8][8] = {
```

Comment:
What's each row and column?

Reply:
Also, since all you need is 8x8 you can use bits instead in `char`.

Reply:
`transitions[FROM][TO]` gives whether or not the transition from state `FROM` to state `TO` is allowed. I'll change it to bits.

Reply:
What I meant was that each row and column should be documented, e.g. how we document valid transitions in the spec here: https://github.com/nutanix/libvfio-user/blob/master/docs/vfio-user.rst#id107